components/device-mgt/org.wso2.carbon.device.mgt.core/src/test/java/org/wso2/carbon/device/mgt/core/cache/DeviceCacheManagerImplTest.java

```diff
@@ -0,0 +1,130 @@
+    private static final String DEVICE_ID_PREFIX = "TEST-DEVICE-ID-";
+    private DeviceCacheManagerImpl deviceCacheManager;
+    private List<DeviceIdentifier> deviceIds = new ArrayList<>();
+    private List<Device> devices = new ArrayList<>();
+    private List<DeviceCacheKey> deviceCacheKeyList = new ArrayList<>();
+
+    @BeforeClass
+    public void init() throws DeviceManagementException, IOException {
+        DeviceConfigurationManager.getInstance().initConfig();
+        for (int i = 0; i < NO_OF_DEVICES; i++) {
+            deviceIds.add(new DeviceIdentifier(DEVICE_ID_PREFIX + i, DEVICE_TYPE));
+        }
+        devices = TestDataHolder.generateDummyDeviceData(this.deviceIds);
+        deviceCacheManager = Mockito.mock(DeviceCacheManagerImpl.class, Mockito.CALLS_REAL_METHODS);
+        DeviceCacheConfiguration configuration = new DeviceCacheConfiguration();
+        configuration.setEnabled(true);
+        DeviceManagementConfig config = DeviceConfigurationManager.getInstance().getDeviceManagementConfig();
+        config.setDeviceCacheConfiguration(configuration);
+        initializeCarbonContext();
+    }
+
+    private void initializeCarbonContext() throws IOException {
+        if (System.getProperty("carbon.home") == null) {
+            File file = new File("src/test/resources");
+            if (file.exists()) {
+                System.setProperty("carbon.home", file.getAbsolutePath());
+            }
+        }
+        PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain(
+                org.wso2.carbon.base.MultitenantConstants.SUPER_TENANT_DOMAIN_NAME);
+        PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantId(
+                org.wso2.carbon.base.MultitenantConstants.SUPER_TENANT_ID);
+    }
+
+
+    @Test(description = "Adding all test devices to the cache")
+    public void testAddDeviceToCache() throws DeviceManagementException {
+        for (int i = 0; i < NO_OF_DEVICES; i++) {
+            deviceCacheManager.addDeviceToCache(deviceIds.get(i), devices.get(i), MultitenantConstants.SUPER_TENANT_ID);
+        }
+    }
+
+    @Test(description = "Adding existing device again", dependsOnMethods = {"testAddDeviceToCache"})
+    public void testAddExistingDeviceToCache() throws DeviceManagementException {
+        deviceCacheManager.addDeviceToCache(deviceIds.get(0), devices.get(0), MultitenantConstants.SUPER_TENANT_ID);
+    }
+
+    @Test(description = "test updating and getting a device in Cache",
+            dependsOnMethods = {"testAddExistingDeviceToCache"})
+    public void testUpdateDeviceInCache() {
+        devices.get(0).setName(UPDATE_NAME);
+        deviceCacheManager.updateDeviceInCache(deviceIds.get(0), devices.get(0), MultitenantConstants.SUPER_TENANT_ID);
+        Device tempDevice = deviceCacheManager.getDeviceFromCache(deviceIds.get(0), MultitenantConstants
+                .SUPER_TENANT_ID);
+        Assert.assertEquals(tempDevice.getName(), UPDATE_NAME);
```

Comment:
Shall we add  another parameter for specifying the reason  in test cases? Shall we add in other places as well